src/erasure-code/clay/ErasureCodeClay.cc

```diff
@@ -0,0 +1,1417 @@
+    err = -EINVAL;
+  }
+  err |= sanity_check_k(k, ss);
+
+  //err |= ErasureCodeJerasure::parse(profile, ss);
+  err |= to_int("d", profile, &d, std::to_string(k+m-1), ss);
+  
+  //*ss << "parse of cl_msr value of d is " << d << std::endl;
+
+  //check for mds block input
+  if (profile.find("mds_type") == profile.end() ||
+      profile.find("mds_type")->second.size() == 0){
+    mds_block = VANDERMONDE_RS;
+    *ss << "mds_type not found in profile" << std::endl;
+  } else {
+    std::string p = profile.find("mds_type")->second;
+    *ss << "recieved mds_type as " << p << std::endl;
+
+    if(p == "reed_sol_van") mds_block = VANDERMONDE_RS;
+    else if(p == "cauchy_mds") mds_block = CAUCHY_MDS;
+    else{
+      mds_block = VANDERMONDE_RS;
+      *ss << "mds_type " << p << "is not currently supported using the default reed_sol_van" << std::endl;
+    }
+
+  }
+   
+  //*ss  << "finished checking mds_type" << std::endl;
+  if (w != 8 ){//&& w != 16 && w != 32) {
+    *ss << "CLMSR: w=" << w
+	<< " must be one of {8, 16, 32} : revert to " << DEFAULT_W << std::endl;
+    profile["w"] = "8";
+    err |= to_int("w", profile, &w, DEFAULT_W, ss);
+    err = -EINVAL;
+    return err;
+  }
+
+
+
+  if ((d < k) || (d > k+m-1)){
+    *ss << "value of d " << d
+	<< " must be within [ " << k << "," << k+m-1 << "]" << std::endl;
+    err = -EINVAL;
+    return err;
+  }
+  q = d-k+1;//this will be m when d=(n-1)=k+m-1
+
+  if( (k+m)%q ){
+    nu = q - (k+m)%q;
+    *ss << "CLMSR: (k+m)%q=" << (k+m)%q
+	 << "q doesn't divide k+m, to use shortening" << std::endl;
+  } else {
+    nu = 0;
+  }
+
+  if(k+m+nu > 254){
+    err = -EINVAL;
+    return err;
+  }
+
+  err |= to_bool("jerasure-per-chunk-alignment", profile,
+		 &per_chunk_alignment, "false", ss);
+  return err;
+}
+
+void ErasureCodeClay::prepare()
+{
+  dout(10) << __func__ << " k:" << k << " m: " << m << " w:" << w << dendl;
+  
+  //get the value for gamma needed for gamma transform
+  int u;
+  for(u = 2; ; u++){
+    if(galois_single_multiply(u,u,w) != 1){
+      gamma = u;
+      break;
+    }
+  }
+
+  //get the matrix needed based on mds_block;
+  switch(mds_block){
+    case VANDERMONDE_RS:
+      matrix = reed_sol_vandermonde_coding_matrix(k+nu, m, w);
+      break;  
+    case CAUCHY_MDS:
+      matrix = cauchy_original_coding_matrix(k+nu, m, w);
+      break;
+    default:
+      dout(0) << __func__ << "mds_block " << mds_block << " not supported" << dendl;
+      assert(0);
+  }
+
+  t = (k+m+nu)/q;
+  sub_chunk_no = pow_int(q, t);
+  dout(10) << __func__ << " q:" << q << " nu:" << nu << " t:" << t << dendl;
+
+  //create B_buf here
+  B_buf = talloc(char*, q*t);
+  assert(B_buf);
+  for(int i = 0; i < q*t; i++)B_buf[i] = NULL;
+}
+
+int ErasureCodeClay::minimum_to_decode(const set<int> &want_to_read,
+			  const set<int> &available,
+			  map<int, vector<pair<int, int>>> *minimum){
+  set<int> minimum_shard_ids;
+
+  if(is_repair(want_to_read, available)) {
```

Comment:
a lot of style inconsistencies in this file:
- space after 'if'
- space between operators and operands like * and %
- lots of commented out bits of dead code

Reply:
Hi Sage,

I am still working on this code. In our previous version, we had clay code as a technique with jerasure plugin. Now we are changing it to a plugin. So it will take us 1-2 weeks to get this code right and working.
I will let you and Kefu know once we have all these changes done.

Thanks,
Myna